expression/builtin_op.go

```diff
@@ -391,6 +407,203 @@ func (b *builtinUnaryOpSig) eval(row []types.Datum) (d types.Datum, err error) {
 	return
 }
 
+type unaryMinusFunctionClass struct {
+	baseFunctionClass
+}
+
+func (b *unaryMinusFunctionClass) typeInfer(argExpr Expression, bf *baseBuiltinFunc) bool {
+	bf.tp.Init(mysql.TypeLonglong)
+	switch argExpr.GetType().Tp {
+	case mysql.TypeString, mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeDouble, mysql.TypeFloat,
+		mysql.TypeDatetime, mysql.TypeDuration, mysql.TypeTimestamp:
+		bf.tp.Tp = mysql.TypeDouble
+	case mysql.TypeNewDecimal:
+		bf.tp.Tp = mysql.TypeNewDecimal
+	}
+	types.SetBinChsClnFlag(bf.tp)
+
+	overflow := false
+	if arg, ok := argExpr.(*Constant); ok {
```

Comment:
@XuHuaiyu, because overflow will be handled only if the statement is Select, and if the args type is not constant, MySQL will just throw a overflow error out, see below:

```
mysql> select * from t;
Field   1:  `a`
Catalog:    `def`
Database:   `test`
Table:      `t`
Org_table:  `t`
Type:       LONGLONG
Collation:  binary (63)
Length:     20
Max_length: 19
Decimals:   0
Flags:      UNSIGNED NUM


+---------------------+
| a                   |
+---------------------+
| 9223372036854775809 |
| 9223372036854775808 |
+---------------------+
2 rows in set (0.00 sec)

mysql> select -a from t;
ERROR 1690 (22003): BIGINT value is out of range in ‘-(`test`.`t`.`a`)’
```